ios/brave-ios/Sources/BraveWallet/WalletUserAssetManager.swift

```diff
@@ -139,32 +226,179 @@ public class WalletUserAssetManager: WalletUserAssetManagerType {
         DataController.performOnMainContext { context in
           let newCoins = self.allNewCoinsIntroduced(networks: allNetworks, context: context)
           if !newCoins.isEmpty {
-            self.migrateUserAssets(for: newCoins, completion: completion)
+            // new coin type introduced, so we want to fetch user assets balances and cache them after
+            // new coin type assets have been migrated to CD
+            self.migrateUserAssets(for: newCoins, completion: { [weak self] in
+              self?.refreshBalances()
+              completion?()
+            })
           } else {
+            // no migration happens. refreshing user assets balance will happen after unlock
             completion?()
           }
         }
       }
     }
   }
   
+  public func getBalances(for asset: BraveWallet.BlockchainToken?, account: String?) -> [WalletUserAssetBalance]? {
+    WalletUserAssetBalance.getBalances(for: asset, account: account)
+  }
+  
+  public func updateBalance(for asset: BraveWallet.BlockchainToken, account: String, balance: String, completion: (() -> Void)?) {
+    WalletUserAssetBalance.updateBalance(
+      for: asset,
+      balance: balance,
+      account: account,
+      completion: completion
+    )
+  }
+  
+  public func removeBalance(for asset: BraveWallet.BlockchainToken, completion: (() -> Void)?) {
+    WalletUserAssetBalance.removeBalance(
+      for: asset,
+      completion: { [weak self] in
+        self?.retriveAllDataObserver().forEach { $0.cachedBalanceRefreshed() }
+        completion?()
+      }
+    )
+  }
+  
+  public func removeBalance(for network: BraveWallet.NetworkInfo, completion: (() -> Void)?) {
+    WalletUserAssetBalance.removeBalance(
+      for: network,
+      completion: { [weak self] in
+        self?.retriveAllDataObserver().forEach { $0.cachedBalanceRefreshed() }
+        completion?()
+      })
+  }
+  
+  public func addUserAssetDataObserver(_ observer: WalletUserAssetDataObserver) {
+    dataObservers.add(observer)
+  }
+  
+  private var refreshBalanceTask: Task<Void, Never>?
+  public func refreshBalances(_ completion: (() -> Void)? = nil) {
+    refreshBalanceTask?.cancel()
+    refreshBalanceTask = Task { @MainActor in
+      let accounts = await keyringService.allAccounts().accounts
+      let allNetworks = await rpcService.allNetworksForSupportedCoins()
+      let allUserAssets: [NetworkAssets] = self.getAllUserAssetsInNetworkAssets(
+        networks: allNetworks,
+        includingUserDeleted: false
+      )
+      await withTaskGroup(
+        of: Void.self,
+        body: { @MainActor [rpcService] group in
+          for account in accounts {
+            group.addTask { @MainActor in
+              let allTokenBalanceForAccount = await rpcService.fetchBalancesForTokens(account: account, networkAssets: allUserAssets)
+              for tokenBalanceForAccount in allTokenBalanceForAccount {
```

Comment:
I believe we're missing a cancelled check (`Task.isCancelled`) in here, otherwise canceling the task at start of function won't stop execution. Not sure if the completion closure should be executed on a cancel, but may pose an issue?